quickjs-libc.c

```diff
@@ -3910,54 +3910,53 @@ JSModuleDef *js_init_module_os(JSContext *ctx, const char *module_name)
 
 /**********************************************************/
 
-#ifdef _WIN32
-static JSValue js_print(JSContext *ctx, JSValue this_val,
-                        int argc, JSValue *argv)
-{
-    int i;
-    const char *str;
-    size_t len;
-    DWORD written;
-    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
-    if (hConsole == INVALID_HANDLE_VALUE)
-        return JS_EXCEPTION;
-    for(i = 0; i < argc; i++) {
-        if (i != 0)
-            WriteConsoleW(hConsole, L" ", 1, &written, NULL);
-        str = JS_ToCStringLen(ctx, &len, argv[i]);
-        if (!str)
-            return JS_EXCEPTION;
-        DWORD prev = GetConsoleOutputCP();
-        SetConsoleOutputCP(CP_UTF8);
-        WriteConsoleA(hConsole, str, len, &written, NULL);
-        SetConsoleOutputCP(prev);
-        JS_FreeCString(ctx, str);
-    }
-    WriteConsoleW(hConsole, L"\n", 1, &written, NULL);
-    FlushFileBuffers(hConsole);
-    return JS_UNDEFINED;
-}
-#else
 static JSValue js_print(JSContext *ctx, JSValue this_val,
                         int argc, JSValue *argv)
 {
+#ifdef _WIN32
+    HANDLE handle;
+    DWORD mode;
+#endif
+    const char *s;
+    DynBuf b;
     int i;
-    const char *str;
-    size_t len;
+
+    dbuf_init(&b);
     for(i = 0; i < argc; i++) {
-        if (i != 0) 
-            putchar(' ');
-        str = JS_ToCStringLen(ctx, &len, argv[i]);
-        if (!str)
-            return JS_EXCEPTION;
-        fwrite(str, 1, len, stdout);
-        JS_FreeCString(ctx, str);
+        s = JS_ToCString(ctx, argv[i]);
+        if (s) {
+            dbuf_printf(&b, "%s%s", &" "[!i], s);
+            JS_FreeCString(ctx, s);
+        } else {
+            dbuf_printf(&b, "%s<exception>", &" "[!i]);
+            JS_FreeValue(ctx, JS_GetException(ctx));
+        }
     }
-    putchar('\n');
+    dbuf_putc(&b, '\n');
+#ifdef _WIN32
+    handle = (HANDLE)_get_osfhandle(/*STDOUT_FILENO*/1); // don't CloseHandle
+    if (GetFileType(handle) != FILE_TYPE_CHAR)
+        goto fallback;
+    if (!GetConsoleMode(handle, &mode))
+        goto fallback;
+    handle = GetStdHandle(STD_OUTPUT_HANDLE);
+    if (handle == INVALID_HANDLE_VALUE)
+        goto fallback;
+    mode = GetConsoleOutputCP();
+    SetConsoleOutputCP(CP_UTF8);
+    WriteConsoleA(handle, b.buf, b.size, NULL, NULL);
+    SetConsoleOutputCP(mode);
+    FlushFileBuffers(handle);
+    goto done;
+fallback:
+#endif
```

Comment:
For casual readers of this code that are unfamiliar with legacy system intricacies, could you add a comment explaining why we need to special case console output on Windows?

I wish we used a more general approach such as detecting the condition once at startup and setting up the system handle `1` or the `stdout` stream to have the expected behavior for all output calls. Hacking this behavior in `js_print` seems incorrect, we should have a lower level API that wraps all output to `stdout`.

Reply:
I'll add that comment.

> detecting the condition once at startup

That's worse because it doesn't handle runtime changes (and from 15 years of node and libuv bug reports I can tell you people _do_ do stuff like reopening stdout at runtime)

Reply:
OK, another potential issue: shouldn't you move the call to `FlushFileBuffers(handle)` before restoring the console mode with `SetConsoleOutputCP(mode)`?  flushing the buffers before the mode change might also be necessary.

This hack really stinks: what if QuickJS is run in a different shell (like cygwin of the old days) or via a remote connection to a Windows host?

Reply:
> shouldn't you move the call to FlushFileBuffers(handle) before restoring the console mode with SetConsoleOutputCP(mode)?

I don't think it matters but I wouldn't mind being proven wrong.

> what if QuickJS is run in a different shell (like cygwin of the old days) or via a remote connection to a Windows host?

We'll just wait for the bug reports to come in :-)